plugins/synaptics-cape/fu-synaptics-cape-firmware.c

```diff
@@ -0,0 +1,215 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2021 Synaptics Incorporated <simon.ho@synaptics.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include <string.h>
+
+#include "fu-synaptics-cape-firmware.h"
+
+typedef struct __attribute__((packed)) {
+	guint32 data[8];
+} FuCapeHidFwCmdUpdateWritePar;
+
+struct _FuSynapticsCapeFirmware {
+	FuSrecFirmwareClass parent_instance;
+	gchar *version;
+	guint16 version_w;
+	guint16 version_x;
+	guint16 version_y;
+	guint16 version_z;
+	guint16 vid;
+	guint16 pid;
+};
+
+G_DEFINE_TYPE(FuSynapticsCapeFirmware, fu_synaptics_cape_firmware, FU_TYPE_FIRMWARE)
+
+guint16
+fu_synaptics_cape_firmware_get_vid(FuSynapticsCapeFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_SYNAPTICS_CAPE_FIRMWARE(self), 0);
+	return self->vid;
+}
+
+guint16
+fu_synaptics_cape_firmware_get_pid(FuSynapticsCapeFirmware *self)
+{
+	g_return_val_if_fail(FU_IS_SYNAPTICS_CAPE_FIRMWARE(self), 0);
+	return self->pid;
+}
+
+static void
+fu_synaptics_cape_firmware_export(FuFirmware *firmware,
+				  FuFirmwareExportFlags flags,
+				  XbBuilderNode *bn)
+{
+	FuSynapticsCapeFirmware *self = FU_SYNAPTICS_CAPE_FIRMWARE(firmware);
+
+	fu_xmlb_builder_insert_kv(bn, "version", self->version);
+	fu_xmlb_builder_insert_kx(bn, "vid", self->vid);
+	fu_xmlb_builder_insert_kx(bn, "pid", self->pid);
+}
+
+static gboolean
+fu_synaptics_cape_firmware_parse_header(FuSynapticsCapeFirmware *self,
+					FuFirmware *firmware,
+					GBytes *fw,
+					GError **error)
+{
+	gsize bufsz = 0x0;
+	const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+	g_autoptr(GBytes) fw_hdr = NULL;
+	g_autoptr(FuFirmware) img_hdr = fu_firmware_new();
+
+	/*the input fw image size shoud be the same as header size */
+	if (bufsz < sizeof(FuCapeHidFileHeader)) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "not enough data to parse header");
+		return FALSE;
+	}
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VID,
+					&self->vid,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_PID,
+					&self->pid,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_X,
+					&self->version_x,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_Y,
+					&self->version_y,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	if (!fu_common_read_uint16_safe(buf,
+					bufsz,
+					FW_CAPE_HID_HEADER_OFFSET_VER_Z,
+					&self->version_z,
+					G_LITTLE_ENDIAN,
+					error))
+		return FALSE;
+
+	self->version = g_strdup_printf("%u.%u.%u.%u",
```

Comment:
again, why save locally?

Reply:
okay, will remove this.

Reply:
well, not remove, just use fu_firmware_set_version() alone.

Reply:
okay